rir/src/compiler/opt/type_speculation.cpp

```diff
@@ -43,12 +43,21 @@ void TypeSpeculation::apply(RirCompiler&, ClosureVersion* function,
                 if (!type.isVoid() && !i->type.isA(type)) {
                     if (auto cp = checkpoint.next(i)) {
                         if (!type.maybeObj()) {
-                            if (type.isA(RType::integer) ||
-                                type.isA(RType::real)) {
-                                speculate[cp][i] = type;
-                            } else {
-                                speculate[cp][i] = i->type.notObject();
-                            }
+                            PirType specType = (type.isA(RType::integer) ||
+                                                type.isA(RType::real))
+                                                   ? type
+                                                   : i->type.notObject();
+                            bool redundant = !Visitor::check(
```

Comment:
we absolutely want to address the actual bug and not just put some ad-hoc test there that prevents it from popping up.

I am not quite sure I understand the problem.

Is it that we speculate twice on the same value? then maybe the fix is to just remove the typefeedback once we speculated. e.g. on line 87 say `typefeedback = PirType::any()`

Reply:
There are 2 bugs. One is a problem in the CFG, where we replace one of the instructions with the speculated type in a BB that isn't preceded by the cast, so the verifier fails with "instruction used before it's defined". That isn't addressed here.

There's also redundant type-speculation, which is fixed by this. This doesn't cause any unsoundness itself, just redundancy. Although if I fix this bug it sidesteps the first bug.

I definitely think we need to address the first bug, and I can look at it later, just not in this PR.

Reply:
ok, but please solve it by not scanning the whole instruction stream. this has like a O(n^2) complexity. removing the feedback should be ok. or, since you are iterating the whole instruction stream here anyways, it could also be just removed again, once we arrive at the redundancy anyway.